src/datasets/iterable_dataset.py

```diff
@@ -355,9 +366,13 @@ def shuffle_data_sources(
     def n_shards(self) -> int:
         return 1
 
-    def shard_data_sources(self, shard_idx: int) -> "HorizontallyConcatenatedMultiSourcesExamplesIterable":
+    def shard_data_sources(
+        self, worker_id: int, num_workers: int
+    ) -> "HorizontallyConcatenatedMultiSourcesExamplesIterable":
         """Either keep only the requested shard, or propagate the request to the underlying iterable."""
-        raise NotImplementedError("Sharding a HorizontallyConcatenatedMultiSourcesExamplesIterable is not implemented")
+        return HorizontallyConcatenatedMultiSourcesExamplesIterable(
+            [iterable.shard_data_sources(worker_id, num_workers) for iterable in self.ex_iterables]
+        )
```

Comment:
I think it should still be NotImplemented for HorizontallyConcatenatedMultiSourcesExamplesIterable, otherwise you could end up with subiterables that are not aligned and don't have the same length

```suggestion
        raise NotImplementedError("Sharding a HorizontallyConcatenatedMultiSourcesExamplesIterable is not implemented")
```